app/forms/floating-ip-edit.tsx

```diff
@@ -0,0 +1,127 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Copyright Oxide Computer Company
+ */
+import { useNavigate, type LoaderFunctionArgs } from 'react-router-dom'
+
+import {
+  apiQueryClient,
+  useApiMutation,
+  useApiQueryClient,
+  usePrefetchedApiQuery,
+} from '@oxide/api'
+import {
+  Listbox,
+  Networking16Icon,
+  PropertiesTable,
+  ResourceLabel,
+  Truncate,
+} from '@oxide/ui'
+import { formatDateTime } from '@oxide/util'
+
+import { SideModalForm } from 'app/components/form'
+import { getFloatingIpSelector, useFloatingIpSelector, useForm } from 'app/hooks'
+import { addToast } from 'app/stores/toast'
+import { pb } from 'app/util/path-builder'
+
+EditFloatingIpSideModalForm.loader = async ({ params }: LoaderFunctionArgs) => {
+  const { project, floatingIp } = getFloatingIpSelector(params)
+  await Promise.all([
+    apiQueryClient.prefetchQuery('floatingIpView', {
+      path: { floatingIp },
+      query: { project },
+    }),
+    apiQueryClient.prefetchQuery('instanceList', {
+      query: { project },
+    }),
+    apiQueryClient.prefetchQuery('floatingIpList', {
+      query: { project },
+    }),
+  ])
+  return null
+}
+
+export function EditFloatingIpSideModalForm() {
+  const { project, floatingIp: floatingIpName } = useFloatingIpSelector()
+  const { data: floatingIp } = usePrefetchedApiQuery('floatingIpView', {
+    path: { floatingIp: floatingIpName },
+    query: { project },
+  })
+  const { data: instances } = usePrefetchedApiQuery('instanceList', {
+    query: { project },
+  })
+  const { data: floatingIps } = usePrefetchedApiQuery('floatingIpList', {
+    query: { project },
+  })
+  console.log(instances, floatingIps)
+
+  const dismissLink = pb.floatingIps({ project })
+  const navigate = useNavigate()
+  const form = useForm({ defaultValues: { instanceId: floatingIp.instanceId } })
+  const onDismiss = () => navigate(dismissLink)
+
+  const queryClient = useApiQueryClient()
+
+  const updateAttachment = useApiMutation('floatingIpAttach', {
+    onSuccess() {
+      addToast({ content: 'Floating IP attached' })
+      queryClient.invalidateQueries('floatingIpView')
+      queryClient.invalidateQueries('floatingIpList')
+      onDismiss()
+    },
+  })
+
+  return (
+    <SideModalForm
+      id="edit-floating-ip-form"
+      form={form}
+      title="Floating IP"
+      onSubmit={(values) => {
+        updateAttachment.mutate({
+          path: { floatingIp: floatingIpName },
+          query: { project },
+          body: {
+            kind: 'instance',
+            parent: values.instanceId,
```

Comment:
If you can be confident that at runtime you can't get here without this being defined (for example by enforcing that instanceis a required field) then the solution would be to use non-null assertion to shut TS up. Whenever I use one I like to include a comment about why. Example:

https://github.com/oxidecomputer/console/blob/b454cefd9b480bd95220b887b4c58d384c0dba94/app/components/form/fields/TlsCertsField.tsx#L69-L78

Another approach would be to use `invariant` to tell TS it's non-null, which I suppose has the advantage of blowing up at runtime if it's violated. I think we used to do that more and stopped.

The most strictly correct way of doing this, which we are not ready to get into, would be to use a `resolver` on the form, which would run the input `FormValues` type through a parse/transform step, in which case I _think_ `onSubmit` only gets called if it succeeds, and it gets the transformed type, so it would no longer worry `instanceId` can be undefined.